src/model/consent/consent.ts

```diff
@@ -0,0 +1,105 @@
+ * which will be addressed in a future ticket
+ */
+
+/*****
+ License
+ --------------
+ Copyright © 2017 Bill & Melinda Gates Foundation
+ The Mojaloop files are made available by the Bill & Melinda Gates Foundation under the Apache License, Version 2.0 (the 'License') and you may not use these files except in compliance with the License. You may obtain a copy of the License at
+ http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, the Mojaloop files are distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ Contributors
+ --------------
+ This is the official list of the Mojaloop project contributors for this file.
+ Names of the original copyright holders (individuals or organizations)
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ - Raman Mangla <ramanmangla@google.com>
+ --------------
+ ******/
+
+import Knex from 'knex'
+
+/*
+ * Interface for Consent resource type
+ */
+export interface Consent {
+  id: string;
+  initiatorId: string;
+  participantId: string;
+  credentialId?: string;
+  credentialType?: string;
+  credentialStatus?: string;
+  credentialPayload?: string;
+  credentialChallenge?: string;
+  createdAt?: string;
+}
+
+/*
+ * Class to abstract Consent DB operations
+ */
+export class ConsentDB {
+  // Knex instance
+  private Db: Knex
+  // Nullable fields which don't allow for explicit null upsert
+  private static nullProtectedProps: Map<string, boolean> = new Map([
+    ['credentialId', true],
+    ['credentialType', true],
+    ['credentialStatus', true],
+    ['credentialPayload', true],
+    ['credentialChallenge', true]
+  ])
+
+  public constructor (dbInstance: Knex) {
+    this.Db = dbInstance
+  }
+
+  // Add initial Consent parameters
+  public register (consent: Consent): Promise<Consent[]> {
+    return this.Db<Consent>('Consent')
+      .insert(consent)
+  }
+
+  // Update Consent credential
+  // Only non-null Consent fields are updated
+  public updateCredentials (consent: Consent): Promise<Consent[]> {
+    const validatedConsent = {}
+
+    for (const key in consent) {
+      // Only allow nullProtectedFields with non-null values for update
+      if (ConsentDB.nullProtectedProps.get(key) && Object(consent)[key]) {
+        Object(validatedConsent)[key] = Object(consent)[key]
+      }
+    }
+
+    return this.Db<Consent>('Consent')
```

Comment:
@lewisdaly @jgeewax there is a corresponding test which ensures we do nothing (no update) if there is no related Consent in DB. Shouldn't we throw an error in such a case?

Reply:
Hmm probably yes. I didn't notice this earlier. I think I assumed that doing a `.where()` in knex when the id doesn't exist would throw an error.

Reply:
So my thought here was that the models were just supposed to act as wrappers over the queries and the actual validation was supposed to happen in domains. Just like SQL, Knex doesn't return an error when it doesn't find a specific row to update. Instead, it does nothing and returns. So I was thinking of checking whether the consent actually exists in the domain logic before even trying to update. Any thoughts? @lewisdaly @jgeewax @eoln

Reply:
This is the case with all the models. I'm adhering to the way SQL would work and letting domains handle the validation.

Reply:
OK I think this is a bigger issue of where validation happens. Can you tell me now what happens if you attempt to update a resource that doesn't exist? Will it just fail silently? Throw an error? Create the row that would've been missing?

If Mojaloop has a standard of what we do here, we should follow that. If not, then we should be explicit about what is and is not permitted. Keep in mind here that putting this inside a controller (rather than at a model level) is likely to be problematic unless Knex provides a way to manage transactional semantics.

In other words, if a handler has to do the following, then we have a problem:

1. Check if item exists. If not, return 404.
2. Update item

What if the item exists but is deleted between steps (1) and (2)? What if the item doesn't exist before step (1) but then exists right after?

The second scenario isn't that scary (just retry the request). The first is very scary if our code has "undefined behavior". As a result, we **must** define the behavior of what happens when someone tries to update a row that doesn't exist. And that behavior (in my opinion) should be to throw some sort of error / reject a promise.

Reply:
OK I just checked the [docs](http://knexjs.org/#Builder-update) more thoroughly. If nothing matches, no updates happen and the result includes an empty array.

In other words, we decide whether a failure happened by looking at the result -- if it's empty, that's a 404 Not Found response code. That's a bit unusual, but not a big deal and avoids transactions entirely which is always fun.

Reply:
We definitely need error handling and validation for checking whether a row exists or not. I thought we could implement it in domains but I agree that it's a better idea to have validation in the model itself. For now, I can return a 404 error from the model but consistent error handling and edge cases will be handled in a future error handling ticket as we have discussed earlier.